File: utils/image-to-ascii/src/keyframes.mjs
// keyframes.mjs: the value of each animated setting at each frame.
//
// A track is one setting's keyframes, as [frame, value] pairs. Before the
// first keyframe the value holds at the first value. After the last keyframe
// it holds at the last value. Between two keyframes it moves by the easing.

export const EASINGS = {
  linear: (t) => t,
  // Smoothstep: slow out of one keyframe and slow into the next.
  smooth: (t) => t * t * (3 - 2 * t),
};

/**
 * Parses "0:100,20:250,40:80" into [[0, 100], [20, 250], [40, 80]], sorted by
 * frame. Throws on a malformed pair or a frame used twice.
 */
export const parseTrack = (text) => {
  const pairs = String(text)
    .split(',')
    .map((part) => part.trim())
    .filter(Boolean)
    .map((part) => {
      const [f, v] = part.split(':').map(Number);
      if (!Number.isInteger(f) || f < 0 || !Number.isFinite(v)) {
        throw new Error(`bad keyframe "${part}": use frame:value, e.g. 20:250`);
      }
      return [f, v];
    });
  return normaliseTrack(pairs);
};

/** Accepts {"0": 100, "20": 250} or [[0, 100], [20, 250]]. */
export const trackFromJson = (json) =>
  normaliseTrack(
    Array.isArray(json)
      ? json.map(([f, v]) => [Number(f), Number(v)])
      : Object.entries(json).map(([f, v]) => [Number(f), Number(v)]),
  );

const normaliseTrack = (pairs) => {
  if (pairs.length === 0)
    throw new Error('a track needs at least one keyframe');
  const sorted = [...pairs].sort((a, b) => a[0] - b[0]);
  for (let i = 1; i < sorted.length; i++) {
    if (sorted[i][0] === sorted[i - 1][0]) {
      throw new Error(`frame ${sorted[i][0]} has two keyframes`);
    }
  }
  return sorted;
};

/** The value of a track at a frame. */
export const valueAt = (track, frame, ease = EASINGS.linear) => {
  if (frame <= track[0][0]) return track[0][1];
  const last = track[track.length - 1];
  if (frame >= last[0]) return last[1];
  let i = 1;
  while (track[i][0] < frame) i++;
  const [f0, v0] = track[i - 1];
  const [f1, v1] = track[i];
  return v0 + (v1 - v0) * ease((frame - f0) / (f1 - f0));
};

/** The frame count that shows every keyframe: the last keyframe plus one. */
export const frameCount = (tracks) =>
  Math.max(1, ...Object.values(tracks).map((t) => t[t.length - 1][0] + 1));

File: utils/image-to-ascii/internal/ascii/keyframes.go
package ascii

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// Track is one setting's keyframes, sorted by frame. Before the first
// keyframe the value holds at the first value. After the last keyframe it
// holds at the last value. Between two keyframes it moves by the easing.
type Track [][2]float64

// Easings maps an easing name to its curve.
var Easings = map[string]func(float64) float64{
	"linear": func(t float64) float64 { return t },
	// Smoothstep: slow out of one keyframe and slow into the next.
	"smooth": func(t float64) float64 { return t * t * (3 - 2*t) },
}

// ParseTrack parses "0:100,20:250,40:80".
func ParseTrack(text string) (Track, error) {
	var t Track
	for _, part := range strings.Split(text, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		fv := strings.SplitN(part, ":", 2)
		bad := fmt.Errorf("bad keyframe %q: use frame:value, e.g. 20:250", part)
		if len(fv) != 2 {
			return nil, bad
		}
		f, err1 := strconv.Atoi(fv[0])
		v, err2 := strconv.ParseFloat(fv[1], 64)
		if err1 != nil || err2 != nil || f < 0 || math.IsInf(v, 0) || math.IsNaN(v) {
			return nil, bad
		}
		t = append(t, [2]float64{float64(f), v})
	}
	return normalise(t)
}

// TrackFromJSON accepts {"0": 100, "20": 250} or [[0, 100], [20, 250]].
func TrackFromJSON(v any) (Track, error) {
	var t Track
	switch j := v.(type) {
	case map[string]any:
		for k, val := range j {
			f, err := strconv.Atoi(k)
			n, ok := val.(float64)
			if err != nil || !ok || f < 0 {
				return nil, fmt.Errorf("bad keyframe %q: %v", k, val)
			}
			t = append(t, [2]float64{float64(f), n})
		}
	case []any:
		for _, pair := range j {
			p, ok := pair.([]any)
			if !ok || len(p) != 2 {
				return nil, fmt.Errorf("bad keyframe %v: use [frame, value]", pair)
			}
			f, ok1 := p[0].(float64)
			n, ok2 := p[1].(float64)
			if !ok1 || !ok2 || f < 0 || f != math.Trunc(f) {
				return nil, fmt.Errorf("bad keyframe %v", pair)
			}
			t = append(t, [2]float64{f, n})
		}
	default:
		return nil, fmt.Errorf("keyframes must be an object or a list of pairs")
	}
	return normalise(t)
}

func normalise(t Track) (Track, error) {
	if len(t) == 0 {
		return nil, fmt.Errorf("a track needs at least one keyframe")
	}
	sort.Slice(t, func(i, j int) bool { return t[i][0] < t[j][0] })
	for i := 1; i < len(t); i++ {
		if t[i][0] == t[i-1][0] {
			return nil, fmt.Errorf("frame %v has two keyframes", t[i][0])
		}
	}
	return t, nil
}

// ValueAt returns the value of a track at a frame.
func (t Track) ValueAt(frame float64, ease func(float64) float64) float64 {
	if frame <= t[0][0] {
		return t[0][1]
	}
	last := t[len(t)-1]
	if frame >= last[0] {
		return last[1]
	}
	i := 1
	for t[i][0] < frame {
		i++
	}
	f0, v0 := t[i-1][0], t[i-1][1]
	f1, v1 := t[i][0], t[i][1]
	return v0 + (v1-v0)*ease((frame-f0)/(f1-f0))
}

// FrameCount returns the count that shows every keyframe: the last keyframe
// of any track, plus one.
func FrameCount(tracks map[string]Track) int {
	n := 1
	for _, t := range tracks {
		if c := int(t[len(t)-1][0]) + 1; c > n {
			n = c
		}
	}
	return n
}
